storage/transactional/reference.go

```diff
@@ -0,0 +1,108 @@
+package transactional
+
+import (
+	"fmt"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+)
+
+var ErrRefHasChanged = fmt.Errorf("reference has changed concurrently")
+
+type ReferenceStorage struct {
+	storer.ReferenceStorer
+	temporal storer.ReferenceStorer
+
+	// deleted, remaining references at this maps are going to be deleted when
+	// commit is requested, the entries are added when RemoveReference is called
+	// and deleted if SetReference is called.
+	deleted map[plumbing.ReferenceName]struct{}
```

Comment:
Sounds good.

However, I'm still concerned that the whole approach is not really usable by borges. If that was the goal, it would actually need its own implementation of `Commit`...

Reply:
Yes, the implementation of `Commit` should be ad-hoc for each interface, but I don't think is going to be the current one.